src/TorchSharp/DisposeScope.cs

```diff
@@ -0,0 +1,301 @@
+﻿// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using TorchSharp.Utils;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Keeps track of all disposables that are in the current scope - the dispose scopes can be nested and the
+    /// nesting functionality is mainly managed by DisposeScopeManager.
+    /// </summary>
+    public class DisposeScope : IDisposable
+    {
+        private readonly DisposeScopeManager _disposeScopeManager;
+
+        public DisposeScope(DisposeScopeManager disposeScopeManager)
+        {
+            _disposeScopeManager = disposeScopeManager;
+            if (disposeScopeManager.DisposeScopeStack.Count > 0) {
+                OuterScope = disposeScopeManager.DisposeScopeStack.Peek();
+            }
+        }
+
+        /// <summary>
+        /// The outer scope with relation to this scope.
+        /// </summary>
+        internal DisposeScope OuterScope { get; set; }
```

Comment:
Not sure if TorchSharp has nullable annotations on. Even if it does not, it might make sense to enable them for new files.

This value looks like it should be nullable.

Reply:
True.

Reply:
> I wonder if there's any point in supporting generic IDisposable over just Tensor instances.

The reason I did it this way was because I'm working on batching and data loading, and I'd either have to track batches with usings and disposes, or hook them to a dispose scope and be done with it. I haven't gotten that far yet, but I think it'll work out.

Reply:
On the other hand, it's a smaller issue than the Tensors, so if it was decided to only support tensors, batches could be handled separately.